suprb/optimizer/solution/ga/crossover.py

```diff
@@ -8,27 +8,49 @@
 
 
 class SolutionCrossover(BaseComponent, metaclass=ABCMeta):
-
-    def __init__(self, crossover_rate: float = 0.9):
-        self.crossover_rate = crossover_rate
-
-    def __call__(self, A: Solution, B: Solution, random_state: RandomState) -> Solution:
-        if random_state.random() < self.crossover_rate:
+    def __call__(self, A: Solution, B: Solution, crossover_rate, random_state: RandomState) -> Solution:
+        if random_state.random() < crossover_rate:
             return self._crossover(A=A, B=B, random_state=random_state)
-        else:
-            # Just return the primary parent
-            return A
+
+        return A
 
     @abstractmethod
     def _crossover(self, A: Solution, B: Solution, random_state: RandomState) -> Solution:
         pass
 
+    def saga2_crossover(
+        self,
+        A: Solution,
+        B: Solution,
+        crossover_rate_min,
+        crossover_rate_max,
+        fitness_mean,
+        fitness_min,
+        fitness_max,
+        random_state: RandomState,
+    ) -> Solution:
+        fitness_parents_mean = (A.fitness_ + B.fitness_) / 2
+
+        crossover_rate_diff = crossover_rate_max - crossover_rate_min
+        fitness_diff = abs(fitness_mean - fitness_parents_mean)
+
+        if fitness_mean in {fitness_min, fitness_max}:
+            crossover_rate = crossover_rate_max
+        elif fitness_parents_mean <= fitness_mean:
+            crossover_rate = crossover_rate_max - crossover_rate_diff * (fitness_diff / (fitness_mean - fitness_min))
+        else:
+            crossover_rate = crossover_rate_max - crossover_rate_diff * (fitness_diff / (fitness_max - fitness_mean))
+
+        if random_state.random() < crossover_rate:
+            return self._crossover(A=A, B=B, random_state=random_state)
+
+        return A
+
 
 class NPoint(SolutionCrossover):
     """Cut the genome at N points and alternate the pieces from solution A and B."""
 
-    def __init__(self, crossover_rate: float = 0.9, n: int = 2):
-        super().__init__(crossover_rate=crossover_rate)
+    def __init__(self, n: int = 2):
```

Comment:
why is the superconstructor no longer invoked?

Reply:
Because the `crossover_rate` moved from the constructor into the function call of `_crossover`, which is why there is no point in calling the super constructor if we are not passing parameters